File: ruby/test/wire/specialists_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class SpecialistsWireTest < WireMockTestCase
  def setup
    super

    @client = Sikaru::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_specialists_list_with_wiremock
    test_id = "specialists.list.0"

    @client.specialists.list(
      project_id: "project_id",
      session_id: "session_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "specialists.list.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/execution-sessions/session_id/specialists",
      query_params: nil,
      expected: 1
    )
  end

  def test_specialists_get_with_wiremock
    test_id = "specialists.get.0"

    @client.specialists.get(
      project_id: "project_id",
      session_id: "session_id",
      thread_id: "thread_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "specialists.get.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id",
      query_params: nil,
      expected: 1
    )
  end

  def test_specialists_cancel_with_wiremock
    test_id = "specialists.cancel.0"

    @client.specialists.cancel(
      project_id: "project_id",
      session_id: "session_id",
      thread_id: "thread_id",
      idempotency_key: "idempotency_key",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "specialists.cancel.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id/cancel",
      query_params: nil,
      expected: 1
    )
  end

  def test_specialists_message_with_wiremock
    test_id = "specialists.message.0"

    @client.specialists.message(
      project_id: "project_id",
      session_id: "session_id",
      thread_id: "thread_id",
      idempotency_key: "idempotency_key",
      input: {
        key: "value"
      },
      request_options: {
        additional_headers: {
          "X-Test-Id" => "specialists.message.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id/messages",
      query_params: nil,
      expected: 1
    )
  end
end

File: rust/tests/specialists_test.rs
use sikaru::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_specialists_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .specialists
        .list(&"project_id".to_string(), &"session_id".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/projects/project_id/execution-sessions/session_id/specialists",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_specialists_get_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .specialists
        .get(
            &"project_id".to_string(),
            &"session_id".to_string(),
            &"thread_id".to_string(),
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_specialists_cancel_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .specialists
        .cancel(
            &"project_id".to_string(),
            &"session_id".to_string(),
            &"thread_id".to_string(),
            &SpecialistCancel {
                idempotency_key: "idempotency_key".to_string(),
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id/cancel",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_specialists_message_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .specialists
        .message(
            &"project_id".to_string(),
            &"session_id".to_string(),
            &"thread_id".to_string(),
            &SpecialistMessage {
                idempotency_key: "idempotency_key".to_string(),
                input: HashMap::from([("key".to_string(), serde_json::json!("value"))]),
                delivery_mode: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/execution-sessions/session_id/specialists/thread_id/messages",
        None,
        1,
    )
    .await
    .unwrap();
}
